[SPARK-10956] Common MemoryManager interface for storage and execution

andrewor14 opened the pull request:
This patch introduces a `MemoryManager` that is the central arbiter of how much memory to grant to storage and execution. This patch is primarily concerned only with refactoring while preserving the existing behavior as much as possible.

This is the first step away from the existing rigid separation of storage and execution memory, which has several major drawbacks discussed on the [issue](https://issues.apache.org/jira/browse/SPARK-10956). It is the precursor of a series of patches that will attempt to address those drawbacks.

Commit:
Introduce MemoryManager + StaticMemoryManager

This commit does not represent any change in behavior. The
MemoryManager's introduced are not actually used anywhere yet.
This will come in the next commit.

Commit:
Derive max memory from MemoryManager

This commit refactors BlockManager and ShuffleMemoryManager to
take in a MemoryManager in their constructors, such that the max
memory is derived from the MemoryManager instead of from a Long
that is passed around everywhere.

Note that no memory acquisition or release goes through the
MemoryManager yet. This will come in future commits.

Commit:
Route ShuffleMemoryManager through new interface

All execution memory is now acquired through the new MemoryManager
interface as of this commit. The next step is to do it for storage
memory.

Commit:
Route MemoryStore through new interface

All storage memory is now acquired through the new MemoryManager
interface. This requires non-trivial refactoring due to the fact
that the unrolling and eviction logic are closely intertwined.

Commit:
Limit scope of synchronized blocks to avoid deadlocks

Commit:
Remove ResultWithDroppedBlocks

It was an awkward case class that didn't really do much. Instead
we'll pass a mutable list into the methods to get the dropped
blocks.

Commit:
Add unit test for StaticMemoryManager

Commit:
Minor style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43293/consoleFull) for   PR 9000 at commit [`88faede`](https://github.com/apache/spark/commit/88faede150c83b197d17f0571fa2af2503a76179).

SparkQA commented:
[Test build #43293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43293/console) for   PR 9000 at commit [`88faede`](https://github.com/apache/spark/commit/88faede150c83b197d17f0571fa2af2503a76179).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43293/
Test PASSed.

rxin commented on core/src/main/scala/org/apache/spark/MemoryManager.scala:
```diff
@@ -0,0 +1,100 @@
+ * sorts and aggregations, while storage memory refers to that used for caching and propagating
+ * internal data across the cluster.
+ */
+private[spark] abstract class MemoryManager {
```
can you move this somewhere else that's not at the top level? I'd like to minimize the number of private[spark] that appears in the top level.

JoshRosen replied:
+1 on moving this out of the top-level and into a new package (maybe `spark.memory` ?). This will also provide a convenient way to group the memory-related classes together post-refactoring.

JoshRosen commented on core/src/main/scala/org/apache/spark/MemoryManager.scala:
```diff
@@ -0,0 +1,100 @@
+  def releaseStorageMemory(numBytes: Long): Unit
+
+  /**
+   * Release all storage memory acquired.
```
This releases all storage memory acquired JVM-wide, across all tasks?

andrewor14 replied:
yes, it's called in `MemoryStore#clear`

JoshRosen replied:
Ah. It looks like `BlockStore#clear` is only called during SparkContext shutdown, so I guess it technically doesn't matter if you update the memory accounting there. Doesn't hurt, I guess, but not strictly necessary.

JoshRosen replied:
You know, if `def releaseStorageMemory(numBytes: Long)` doesn't throw when releasing more than the total amount of storage memory, you could just eliminate this method and call `releaseStorageMemory(Long.MAX_VALUE)` instead.

andrewor14 replied:
I think releasing more than you have is a bad sign. I actually log a warning for that. It's probably OK to just keep this method.

JoshRosen commented on core/src/main/scala/org/apache/spark/MemoryManager.scala:
```diff
@@ -0,0 +1,100 @@
+
+
+/**
+ * An abstract memory manager that enforces how memory is shared between execution and storage.
```
Can you add a comment to clarify whether this is a one-per-JVM component or one-per-task component?

JoshRosen commented on core/src/main/scala/org/apache/spark/StaticMemoryManager.scala:
```diff
@@ -0,0 +1,229 @@
+  }
+
+  // For testing only
+  def setMemoryStore(store: MemoryStore): Unit = {
```
Why not just accept the `MemoryStore` via the constructor?

andrewor14 replied:
initialization order :(

JoshRosen replied:
Would it be possible to explicitly call this method in the non-testing path as well? e.g. remove the SparkEnv.get from inside of this class and call this when wiring up the components?

andrewor14 replied:
yeah, in `BlockManager`'s constructor we can do the following:
```
private val memoryStore = new MemoryStore(...)
memoryManager.setMemoryStore(memoryStore)
```
do you think that's better?

JoshRosen replied:
Yes, I think that's clearer.

JoshRosen commented on core/src/main/scala/org/apache/spark/StaticMemoryManager.scala:
```diff
@@ -0,0 +1,229 @@
+
+  // Amount of execution memory in use. Accesses must be synchronized on `executionLock`.
+  private var _executionMemoryUsed: Long = 0
+  private val executionLock = new Object
```
Just curious: do we need two separate locks, or would using a single lock by synchronizing on `this` suffice?

andrewor14 replied:
I just thought we'd have a little more parallelism if we use two locks

JoshRosen commented on core/src/main/scala/org/apache/spark/StaticMemoryManager.scala:
```diff
@@ -0,0 +1,229 @@
+      blockId: BlockId,
+      numBytes: Long,
+      evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Long = {
+    val currentUnrollMemory = memoryStore.currentUnrollMemory
```
Are there any synchronization concerns here w.r.t. `currentUnrollMemory`? Haven't thought this through, but just wanted to ask.

andrewor14 replied:
there's this `accountingLock` in `MemoryStore` that you always acquire before requesting any storage / unroll memory, so it should already be handled there. Maybe we should add a comment there to express this assumption.

andrewor14 replied:
added in `reserveUnrollMemoryForThisTask`

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/MemoryStore.scala:
```diff
@@ -409,40 +397,42 @@ private[spark] class MemoryStore(blockManager: BlockManager, maxMemory: Long)
    * from the same RDD (which leads to a wasteful cyclic replacement pattern for RDDs that
    * don't fit into memory that we want to avoid).
    *
-   * Assume that `accountingLock` is held by the caller to ensure only one thread is dropping
```
Did you mean to omit this part of the comment?

andrewor14 replied:
yes, `ensureFreeSpace` now explicitly holds the `accountingLock`, since it can be called from random places like our memory manager

JoshRosen commented on core/src/test/scala/org/apache/spark/StaticMemoryManagerSuite.scala:
```diff
@@ -0,0 +1,169 @@
+      blockId: BlockId,
+      numBytes: Long): Unit = {
+    verify(ms).ensureFreeSpace(meq(blockId), meq(numBytes: java.lang.Long), any())
+    reset(ms)
```
`reset` is usually considered an anti-pattern in Mockito, but I guess it's okay here because although it forgets interactions _and_ stubbing you haven't stubbed any methods on `mm`.

andrewor14 replied:
yeah, it actually calls `ensureFreeSpace(1)` 3 times in total, but I need a way to assert that it's called exactly once after each acquire. I couldn't find another way to do this.

JoshRosen commented on core/src/test/scala/org/apache/spark/StaticMemoryManagerSuite.scala:
```diff
@@ -0,0 +1,169 @@
+    assert(mm.acquireExecutionMemory(1L) === 1L)
+    assert(mm.executionMemoryUsed === 201L)
+    // Release beyond what was acquired
+    mm.releaseExecutionMemory(maxExecutionMem)
```
Do you think that releasing beyond what was acquired should throw an exception, since this may indicate an accounting error?

andrewor14 replied:
it logs a warning

JoshRosen commented on core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala:
```diff
@@ -138,30 +150,27 @@ class ShuffleMemoryManager protected (
   def create(maxMemory: Long, pageSizeBytes: Long): ShuffleMemoryManager = {
-    new ShuffleMemoryManager(maxMemory, pageSizeBytes)
+    val conf = new SparkConf
+    val memoryManager = new StaticMemoryManager(conf, maxMemory, Long.MaxValue)
```
This is a totally minor nit to be picking, but do you mind explicitly naming these parameters at their call sites?  e.g.

```
new StaticMemoryManager(conf, maxExecutionMemory = maxMemory, maxStorageMemory = Long.MaxValue)
```

I think this makes the code slightly easier to read.

JoshRosen commented on core/src/test/scala/org/apache/spark/StaticMemoryManagerSuite.scala:
```diff
@@ -0,0 +1,169 @@
+    assert(mm.acquireExecutionMemory(1000L) === 890L)
+    assert(mm.executionMemoryUsed === 1000L)
+    assert(mm.acquireExecutionMemory(1L) === 0L)
+    assert(mm.executionMemoryUsed === 1000L)
```
Could use 

```
assert(mm.executionMemoryUsed === maxExecutionMem)
````

to make this marginally easier to read.

JoshRosen commented on core/src/test/scala/org/apache/spark/StaticMemoryManagerSuite.scala:
```diff
@@ -0,0 +1,169 @@
+  test("basic storage memory") {
+    val maxStorageMem = 1000L
+    val dummyBlock = TestBlockId("you can see the world you brought to live")
+    val dummyBlocks = new ArrayBuffer[(BlockId, BlockStatus)]
```
Could you rename this to `evictedBlocks`, since that's what it is?

JoshRosen replied:
Also, I noticed that there's only one assert involving `dummyBlocks`. In order to test the case where it's non-empty, I guess we'd have to more realistically stub the MemoryStore.

JoshRosen replied:
However, since you call reset() on the MemoryStore between test cases within this method I suppose that you could stub the ensureFreeSpace() method on MemoryStore to have it return some dummy evicted blocks, letting you assert that the plumbing of `dummyBlocks` is correct. This might be overkill, though.

andrewor14 replied:
meh, let's not do that

JoshRosen commented on core/src/test/scala/org/apache/spark/StaticMemoryManagerSuite.scala:
```diff
@@ -0,0 +1,169 @@
+    assert(mm.storageMemoryUsed === 10L)
+    assert(dummyBlocks.isEmpty)
+    assert(mm.acquireStorageMemory(dummyBlock, 100L, dummyBlocks) === 100L)
+    assertEnsureFreeSpaceCalled(ms, dummyBlock, 100L)
```
Could assert that 110 of storage memory is used here, prior to trying to acquire more in next call.

JoshRosen commented on core/src/test/scala/org/apache/spark/StaticMemoryManagerSuite.scala:
```diff
@@ -0,0 +1,169 @@
+    assert(mm.storageMemoryUsed === 50L)
+    assert(mm.executionMemoryUsed === 67L)
+    // Only storage memory should be released
+    mm.releaseStorageMemory()
```
For completeness, want to also call the variant of `releaseStorageMemory` which takes a specific amount of memory to release?

andrewor14 replied:
this is tested in the previous test

JoshRosen commented on core/src/test/scala/org/apache/spark/StaticMemoryManagerSuite.scala:
```diff
@@ -0,0 +1,169 @@
+    assert(mm.storageMemoryUsed === 60L)
+    when(ms.currentUnrollMemory).thenReturn(60L)
+    assert(mm.acquireUnrollMemory(dummyBlock, 500L, dummyBlocks) === 500L)
+    // `spark.storage.unrollFraction` is 0.4, so the max unroll space is 400 bytes.
```
Is it possible to add an assertion for this on the following line?

andrewor14 replied:
the max unroll space is private so we don't have access to it. I think it's OK to leave it out since the `340L` assert implies that it was `400L`

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkEnv.scala:
```diff
@@ -69,6 +69,8 @@ class SparkEnv (
     val httpFileServer: HttpFileServer,
     val sparkFilesDir: String,
     val metricsSystem: MetricsSystem,
+    // TODO: unify these *MemoryManager classes
```
There's a JIRA for this, so you could mention the number here.

JoshRosen commented on core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:
```diff
@@ -1040,17 +1044,23 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
   test("reserve/release unroll memory") {
     store = makeBlockManager(12000)
     val memoryStore = store.memoryStore
+    val dummyBlock = TestBlockId("")
```
Since these variables are only used within `reserveUnrollMemoryForThisTask` you could define them inside of that method.

andrewor14 replied:
ok, I wanted to avoid making an array buffer every time we call that, but I guess it's not that expensive.

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/MemoryStore.scala:
```diff
@@ -489,17 +478,21 @@ private[spark] class MemoryStore(blockManager: BlockManager, maxMemory: Long)
+      if (success) {
         val taskAttemptId = currentTaskAttemptId()
         unrollMemoryMap(taskAttemptId) = unrollMemoryMap.getOrElse(taskAttemptId, 0L) + memory
       }
```
AFAIK, `acquireUnrollMemory` can return less than the requested number of bytes, so don't you have to free the acquired unroll memory in an `else` branch here?

andrewor14 replied:
as discussed offline, unroll memory returns all or nothing. We'll change the signature to return boolean to make this clear.

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/MemoryStore.scala:
```diff
@@ -507,16 +500,17 @@ private[spark] class MemoryStore(blockManager: BlockManager, maxMemory: Long)
    * If the amount is not specified, remove the current task's allocation altogether.
    */
-  def releaseUnrollMemoryForThisTask(memory: Long = -1L): Unit = {
+  def releaseUnrollMemoryForThisTask(memory: Long = Long.MaxValue): Unit = {
```
Thanks for cleaning this up. The old `-1L` here was really confusing.

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/MemoryStore.scala:
```diff
@@ -507,16 +500,17 @@ private[spark] class MemoryStore(blockManager: BlockManager, maxMemory: Long)
+        val memoryToRelease = math.min(memory, unrollMemoryMap(taskAttemptId))
+        if (memoryToRelease > 0) {
+          unrollMemoryMap(taskAttemptId) -= memoryToRelease
+          if (unrollMemoryMap(taskAttemptId) == 0) {
```
The old code used `<= 0`. Do we also need to worry about the `<` case here?

JoshRosen replied:
Actually, I guess not since we never decrement by more than the current value.

JoshRosen commented on core/src/main/scala/org/apache/spark/storage/MemoryStore.scala:
```diff
@@ -525,22 +519,40 @@ private[spark] class MemoryStore(blockManager: BlockManager, maxMemory: Long)
+      if (success) {
+        pendingUnrollMemoryMap(taskAttemptId) =
+          pendingUnrollMemoryMap.getOrElse(taskAttemptId, 0L) + memory
+      }
```
Similar comment here: do we need to have an `else` branch here to free `acquired` when it's less than `memory`?

JoshRosen replied:
I just noticed that the return value of `reservePendingUnrollMemoryForThisTask` is never checked. Does this mean that we never expect this operation to fail?

JoshRosen replied:
Andrew and I discussed this offline and found a way to clean up the logic here: https://github.com/andrewor14/spark/pull/2

JoshRosen committed:
Add comment to pending unroll.

JoshRosen committed:
Pending unroll transfer.

Commit:
Merge pull request #2 from JoshRosen/pending-unroll-comment

Clarify the transfer of memory from unroll to pending unroll

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43437 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43437/consoleFull) for   PR 9000 at commit [`adb1764`](https://github.com/apache/spark/commit/adb1764c851f3ef0ed2ffd56a3fe9b65330b8abf).

JoshRosen commented:
LGTM, by the way, pending updates to address comments. The majority of the unaddressed comments are minor and nitpicky, but I think that https://github.com/apache/spark/pull/9000#discussion_r41568609 might deserve special attention.

Commit:
Merge branch 'memory-manager' of github.com:andrewor14/spark into memory-manager

Commit:
Merge branch 'master' of github.com:apache/spark into memory-manager

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43445 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43445/consoleFull) for   PR 9000 at commit [`6e2e870`](https://github.com/apache/spark/commit/6e2e870ab5610d6a0152021ec68bb276422ac963).

Commit:
Move MemoryManager to new package

Commit:
Change acquire[Unroll/Storage]Memory to return boolean

Previously these would always return either 0 or N. Instead we
can express this simply as whether the acquire was successful
so downstream usages don't have to worry about releasing partially
acquired memory.

Commit:
Address comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43450 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43450/consoleFull) for   PR 9000 at commit [`fc7f9f5`](https://github.com/apache/spark/commit/fc7f9f519852c2b3ef3eebcbc8e3f0ba63fcb3dc).

andrewor14 commented:
@JoshRosen alright I believe I addressed all of your comments. Please let me know if it LGTY.

JoshRosen commented:
LGTM. Feel free to merge pending Jenkins.

SparkQA commented:
[Test build #43437 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43437/console) for   PR 9000 at commit [`adb1764`](https://github.com/apache/spark/commit/adb1764c851f3ef0ed2ffd56a3fe9b65330b8abf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43437/
Test PASSed.

andrewor14 commented:
Merging

andrewor14 commented:
Never mind, the test pass was not for the latest commit.

SparkQA commented:
[Test build #43445 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43445/console) for   PR 9000 at commit [`6e2e870`](https://github.com/apache/spark/commit/6e2e870ab5610d6a0152021ec68bb276422ac963).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43445/
Test PASSed.

SparkQA commented:
[Test build #43450 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43450/console) for   PR 9000 at commit [`fc7f9f5`](https://github.com/apache/spark/commit/fc7f9f519852c2b3ef3eebcbc8e3f0ba63fcb3dc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43450/
Test PASSed.

JoshRosen commented:
Woohoo! Tests have passed as of the latest commit, so I'm going to merge this now.
